rasa/core/policies/intent_ted_policy.py

```diff
@@ -0,0 +1,492 @@
+        """
+        if self.model is None:
+            return self._prediction(self._default_predictions(domain))
+
+        # create model data from tracker
+        tracker_state_features = self._featurize_for_prediction(
+            tracker, domain, interpreter
+        )
+
+        model_data = self._create_model_data(tracker_state_features)
+        output = self.model.run_inference(model_data)
+
+        # take the last prediction in the sequence
+        similarities = output["similarities"][:, -1, :]
+
+        # Check for unlikely intent
+        is_unlikely_intent = self._check_unlikely_intent(domain, similarities, tracker)
+
+        confidences = list(np.zeros(domain.num_actions))
+
+        if is_unlikely_intent:
+            confidences[domain.index_for_action(ACTION_UNLIKELY_INTENT_NAME)] = 1.0
+
+        return self._prediction(
+            confidences,
+            action_metadata=self._collect_action_metadata(domain, similarities),
+        )
+
+    def _should_check_for_intent(self, intent: Text, domain: Domain) -> bool:
+
+        if domain.intents.index(intent) not in self.intent_thresholds:
+            # This means the intent was never present in a story
+            return False
+        if intent in self.config[IGNORE_INTENTS_LIST]:
+            return False
+
+        return True
+
+    def _check_unlikely_intent(
+        self, domain: Domain, similarities: np.array, tracker: DialogueStateTracker
+    ) -> bool:
+        """Check if the latest user event is probable according to IntentTED predictions.
+
+        If the similarity prediction for the intent of
+        latest user event is lower than the threshold
+        calculated for that intent during training, the
+        corresponding user intent is unlikely.
+
+        Args:
+            domain: Domain of the assistant.
+            similarities: Predicted similarities for all labels.
+            tracker: Current conversation tracker
+        """
+        predicted_intent_scores = {
+            index: similarities[0][index] for index, intent in enumerate(domain.intents)
+        }
+        sorted_intent_scores = sorted(
+            [
+                (intent_label, score)
+                for intent_label, score in predicted_intent_scores.items()
+            ],
+            key=lambda x: x[1],
+        )
+        # Get the last intent prediction from tracker
+        last_user_event: Optional[UserUttered] = tracker.get_last_event_for(UserUttered)
+        if last_user_event:
+            query_label = last_user_event.intent_name
+            query_label_id = domain.intents.index(query_label)
+            query_label_score = similarities[0][query_label_id]
+
+            logger.debug(f"Querying for intent {query_label}")
+
+            if self._should_check_for_intent(query_label, domain):
+
+                logger.debug(
+                    f"Score for user intent {query_label} likely to occur here is "
```

Comment:
Can we say "Probability for intent {query_label} to occur here is, ..."?

Reply:
It's not a probability. It's dot product similarity which is unbounded.

Reply:
Right. But "user intent" is redundant. Just "intent" would do, and we don't need "likely".